src/ReverseProxy/Abstractions/RouteDiscovery/Contract/ProxyMatch.cs

```diff
@@ -65,7 +64,42 @@ internal static bool Equals(ProxyMatch proxyMatch1, ProxyMatch proxyMatch2)
 
             return string.Equals(proxyMatch1.Path, proxyMatch2.Path, StringComparison.OrdinalIgnoreCase)
                 && CaseInsensitiveEqualHelper.Equals(proxyMatch1.Hosts, proxyMatch2.Hosts)
-                && CaseInsensitiveEqualHelper.Equals(proxyMatch1.Methods, proxyMatch2.Methods);
+                && CaseInsensitiveEqualHelper.Equals(proxyMatch1.Methods, proxyMatch2.Methods)
+                && HeadersEqual(proxyMatch1.Headers, proxyMatch2.Headers);
+        }
+
+        // Order sensitive to reduce complexity
+        private static bool HeadersEqual(IReadOnlyList<RouteHeader> headers1, IReadOnlyList<RouteHeader> headers2)
+        {
+            if (ReferenceEquals(headers1, headers2))
+            {
+                return true;
+            }
+
+            if ((headers1?.Count ?? 0) == 0 && (headers2?.Count ?? 0) == 0)
+            {
+                return true;
+            }
+
+            if (headers1 != null && headers2 == null || headers1 == null && headers2 != null)
```

Comment:
It can be simplified because we checked at line 74 that they are not equal, thus if one is null the other is guaranteed to be **not null**.
```suggestion
            if (headers1 == null || headers2 == null)
```